src/ray/core_worker/store_provider/store_provider.h

```diff
@@ -1,36 +1,84 @@
 #ifndef RAY_CORE_WORKER_STORE_PROVIDER_H
 #define RAY_CORE_WORKER_STORE_PROVIDER_H
 
+#include "arrow/python/serialize.h"
+
 #include "ray/common/buffer.h"
 #include "ray/common/id.h"
 #include "ray/common/status.h"
 #include "ray/core_worker/common.h"
 
 namespace ray {
 
-/// Binary representation of ray object.
+const std::string kBufferTooSmallErrMsg = "Target buffer smaller than object data.";
+
+/// RayObjects consist of data and metadata. Metadata is always passed
+/// in as a buffer because it is small, but data can be handled
+/// differently by implementations in order to avoid copying large data.
 class RayObject {
  public:
-  /// Create a ray object instance.
-  ///
-  /// \param[in] data Data of the ray object.
-  /// \param[in] metadata Metadata of the ray object.
-  RayObject(const std::shared_ptr<Buffer> &data, const std::shared_ptr<Buffer> &metadata)
-      : data_(data), metadata_(metadata) {}
+  RayObject(const std::shared_ptr<Buffer> &metadata) : metadata_(metadata) {}
+
+  virtual ~RayObject() {}
 
-  /// Return the data of the ray object.
-  const std::shared_ptr<Buffer> &GetData() const { return data_; };
+  const std::shared_ptr<Buffer> &Metadata() const { return metadata_; };
 
-  /// Return the metadata of the ray object.
-  const std::shared_ptr<Buffer> &GetMetadata() const { return metadata_; };
+  virtual const std::shared_ptr<Buffer> &Data() = 0;
+
+  virtual const size_t DataSize() const = 0;
+
+  virtual Status WriteDataTo(std::shared_ptr<Buffer> buffer) const = 0;
 
  private:
-  /// Data of the ray object.
-  const std::shared_ptr<Buffer> data_;
-  /// Metadata of the ray object.
   const std::shared_ptr<Buffer> metadata_;
 };
 
+/// BufferedRayObject is a RayObject whose data is held in a buffer.
+class BufferedRayObject : public RayObject {
+ public:
+  BufferedRayObject(const std::shared_ptr<Buffer> &metadata,
+                    const std::shared_ptr<Buffer> &data)
+      : RayObject(metadata), data_(data) {}
+
+  const std::shared_ptr<Buffer> &Data() override { return data_; };
+
+  const size_t DataSize() const override { return data_->Size(); };
+
+  Status WriteDataTo(std::shared_ptr<Buffer> buffer) const override;
+
+ private:
+  const std::shared_ptr<Buffer> data_;
+};
+
+/// PyArrowRayObject is a RayObject whose data is a SerializedPyObject passed
+/// in from an Arrow serialization. This is not placed into a buffer because
+/// it consists of a collection of pointers directly into Python memory. The
+/// WriteDataTo method will directly write from this memory. Note that calling
+/// Data() for the first time will induce a copy, because it must copy data
+/// from Python memory into a buffer.
```

Comment:
Maybe it's better to move this to `Data()`'s comment.
```
Note that calling
/// Data() for the first time will induce a copy, because it must copy data
/// from Python memory into a buffer.
```